Update config instructions (#56)
LycheeOrg/LycheeOrg.github.io: docs/installation.md

### Configuration

> After installing Lychee, some additional configuration is necessary as detailed out below.

Reviewer: In general I'm against adding words for the sake of it, but ok.
```suggestion
After installing Lychee, some additional configuration is necessary as detailed below.
```

#### Public Directory

Configure your web server's root to be the `public` directory. The `index.php` in this directory serves as the interface for all HTTP requests to Lychee's API.

#### Directory Permissions

> You need to configure some file and directory permissions. Directories within the `storage` and the `bootstrap/cache` directories should be writable by your web server or Lychee will not run. Additionally, `public/uploads`, `public/dist` and `public/sym` need to be writable to upload files, customise the CSS and utilise [symbolic links](https://lycheeorg.github.io/docs/settings.html#symbolic-link) respectively.

Reviewer: ```suggestion
Directories within the `storage` and the `bootstrap/cache` directories must be writable by your web server. Additionally, `public/uploads`, `public/dist` and `public/sym` need to be writable to upload files, customise the CSS and utilise [symbolic links](https://lycheeorg.github.io/docs/settings.html#symbolic-link) respectively.
```

#### Application Settings

The main directory should contain a `.env` file.

> Normally, Composer has generated this file for you based on `.env.example` while Composer has been installing the dependencies (see above).
> If the file `.env` does not exist, then you should copy `.env.example` to a new file named `.env` now.

Reviewer: ```suggestion
Normally, Composer will have copied this file for you while installing the dependencies.
If the `.env` file does not exist, then you should copy `.env.example` to a new file named `.env` now.
```

I still don't like the second line, but I can't improve it at the moment.

Note that the `.env` file should **not** be exposed online as it contains the encryption key used for cookies as well as database credentials.

##### Application Key

The `.env` file contains the setting `APP_KEY` which holds your application key.

> Use the `php artisan key:generate` command to set the application key to a random value.

Reviewer: This is also done by composer, but repeating it does no harm.

Reviewer 2: I wonder more about the 32 characters. I used this command and in my case the result was a string starting with `base64:...` and then followed by 48 characters. I am not sure what happens, if you only use 32 characters or if you get low entropy, because you cannot express a truly random string with printable characters only.

Do we really want to recommend users to create their own random string?

> If you don't want to use this command and set the key yourself, this value should be 32 characters long.

Reviewer: ```suggestion
If you don't want to use this command and choose to set the key yourself, it should be 32 characters long.
```

Reviewer: *from elsewhere* ....this line also needs revision.

**If the application key is not set, your user sessions and other encrypted data will not be secure!**

##### Additional Configuration

> The settings of the default `.env` file provide usable and operational settings out-of-the-box. 

Reviewer: ```suggestion
The default `.env` file provides usable settings out of the box.
```

You may want to configure a few additional components of Lychee, such as:

- Cache
- Database
- Session

> Some advanced options cannot be configured through the `.env` file, but you can take a look into the directory `config/` to find such options. Usually, you don't have to change them and can go with the defaults. **If you don't know what they're doing, do not change them.**

Reviewer: To me this is more of an invitation than @qwerty287's version. I don't like it.

Reviewer 2: I am not sure if Github is out-of-sync again or if were are looking at different things, but this is @qwerty287's version. (confused)

Reviewer: It's different, but less different than I remembered. Did we decide that only the ones in `app.php` were missing from `.env`?

Other than that.... I can't suggest here so...

### Configuration

After installing Lychee, some additional configuration is necessary as detailed out below.

> #### Public Directory

Reviewer 2: I wonder wether this section actually belongs here or should be moved below `## Webserver configuration`.

Reviewer: > I wonder wether this section actually belongs here or should be moved below `## Webserver configuration`.

Fair point. The `public` directory stuff, anyway.

Reviewer 2: OK, I move the `public` directory stuff.

Reviewer 2: Fixed in https://github.com/LycheeOrg/LycheeOrg.github.io/pull/56/commits/de1be208cc49bbbcb794b7f29764122081be4c7b

Note that the `.env` file should **not** be exposed online as it contains the encryption key used for cookies as well as database credentials.

##### Application Key

The `.env` file contains the setting `APP_KEY` which holds your application key.
Normally, Composer sets this key for you while installing the dependencies.
If it is missing, Lychee will not work. Then, you can use the `php artisan key:generate` command to set the application key to a random value.

##### Additional Configuration

The default `.env` file provides usable settings out of the box.

You may want to configure a few additional components of Lychee, such as:

- Cache
- Database
- Session

They are documented in the `.env.example` file, which may have been copied into your `.env`. They are also listed on the [configuration](configuration.html) page.

##### Advanced Configuration (Dangerous)

Some advanced options cannot be configured through the `.env` file. If you really need to change them, look in the `config/` directory.

**If you don't know what they're doing, do not change them.**

## Web Server Configuration

### General

> Configure your web server's root to be the `public` directory.

Reviewer 3: We've been asked about this recently I believe -- could we make this sentence more... nuanced? Many won't want to or even be able to change the root, so we shouldn't make it sound like this is a hard requirement. How about saying something like:

"Configure your web server to make the `public` directory accessible via a URL. This could be the root of your web server (short and convenient) or, if you prefer, a subfolder such as `photos`".

Reviewer: That could work. If we do it may be worth reiterating that the level above shouldn't be publicly accessible.

Reviewer 2: I'm in strong favor of both comments  Repeating the warning is something we should really do. It was only recently when we had the issue that a user made the root of Lychee publicly accessible and the user tried to access Lychee via putting `public/` in the URL.

Btw: It just crosses my mind that we do not mention that the setting `APP_URL` must be set to the _external_ URL by which the `public/` folder can be reached. This should be part of the configuration of the `.env` file but it won't harm to repeat it here again.

### Configuration

#### Directory Permissions

Directories within the `storage` and the `bootstrap/cache` directories must be writable by your web server or Lychee will not run. Additionally, `public/uploads`, `public/dist` and `public/sym` need to be writable to upload files, customise the CSS and utilise [symbolic links](https://lycheeorg.github.io/docs/settings.html#symbolic-link) respectively.

#### Application Settings

The main directory should contain a `.env` file.
Normally, Composer will have copied this file for you while installing the dependencies.
If the `.env` file does not exist, then you should copy `.env.example` to a new file named `.env` now.
Note that the `.env` file should **not** be exposed online as it contains the encryption key used for cookies as well as database credentials.

##### Application Key

The `.env` file contains the setting `APP_KEY` which holds your application key.
Use the `php artisan key:generate` command to set the application key to a random value.

> **If the application key is not set, your user sessions and other encrypted data will not be secure!**

Reviewer: I forget - will Lychee even run without it set?

Reviewer: I was just wondering whether "will not be secure" is relevant if "Lychee will not work" applies? No strong feelings.

Reviewer: (Just seen that's already updated!)

##### Additional Configuration

The default `.env` file provides usable settings out of the box.

You may want to configure a few additional components of Lychee, such as:

- Cache
- Database
- Session

They are documented in the `.env.example` file, which may have been copied into your `.env`. They are also listed on the [configuration](configuration.html) page.

##### Advanced Configuration (Dangerous)

> Some advanced options cannot be configured through the `.env` file, but you may take a look into the directory `config/` to find such options.
> Usually, you don't have to change them and can go with the defaults.

Reviewer: ... maybe this? I'm not sure, I just feel like it leans harder towards "don't do it" or "here be dragons".
```suggestion
Some advanced options cannot be configured through the `.env` file. If you really need to change them, look in the `config/` directory.
```

#### Public Directory

> After installing Lychee, configure your web server's root to be the `public` directory. The `index.php` in this directory serves as the interface for all HTTP requests to Lychee's API. Note that the `.env` file should **not** be exposed online as it contains the encryption key used for cookies as well as database credentials.

Reviewer: As per other comment...
```suggestion
Configure your web server's root to be the `public` directory. The `index.php` in this directory serves as the interface for all HTTP requests to Lychee's API. Note that the `.env` file should **not** be exposed online as it contains the encryption key used for cookies as well as database credentials.
```

Reviewer 2: Included in my commit.

#### Directory Permissions

> After installing Lychee, you need to configure some permissions. Directories within the `storage` and the `bootstrap/cache` directories should be writable by your web server or Lychee will not run. Additionaly `public/uploads`, `public/dist` and `public/sym` need to be writable to upload files, customise the CSS and utilise [symbolic links](https://lycheeorg.github.io/docs/settings.html#symbolic-link) respectively.

Reviewer 2: The "after installing Lychee" is repeated from some lines above. Again, I can only apologize and give the lame excuse that I have never read the instructions thoroughly. That is what happens when you invite me to a review :)

I would rephrase it like this:

The phrase "after installing..." should only be used once and put directly below the 3rd-level caption `### Configuration`. This also avoids having no text between two consecutive captions. Something like this "After installing Lychee some adjustments must be made as detailed out below". Then remove that phrase from everywhere else, because all of these steps are after the installation.

Reviewer: I'd go further and say that configuration is inherently going to be after installation.

> After installing Lychee, you need to configure some permissions. Directories within the `storage` and the `bootstrap/cache` directories should be writable by your web server or Lychee will not run. Additionaly `public/uploads`, `public/dist` and `public/sym` need to be writable to upload files, customise the CSS and utilise [symbolic links](https://lycheeorg.github.io/docs/settings.html#symbolic-link) respectively.

Reviewer: As per other comment...

```suggestion
Directories within the `storage` and the `bootstrap/cache` directories must be writable by your web server or Lychee will not run. Additionaly `public/uploads`, `public/dist` and `public/sym` need to be writable to upload files, customise the CSS and utilise [symbolic links](https://lycheeorg.github.io/docs/settings.html#symbolic-link) respectively.
```

Reviewer 2: Sorry, we had a concurrency problem here, because I was working on an update and pushed it while you added your suggestion. I added the suggestion to my commit.

##### Application Key

The `.env` file contains the setting `APP_KEY` which holds your application key.
Normally, Composer sets this key for you while installing the dependencies.

> If it is missing, Lychee will not work. Then, you can use the `php artisan key:generate` command to set the application key to a random value.

Reviewer 2: ```suggestion
If the key is not set, Lychee will not work. Then, you can use the `php artisan key:generate` command to set the application key to a random value.
```

If one reads sloppily or in a hurry, one could think that "it" refers to "Composer" and I was wondering at first why Composer should be missing.

(Full Sheldon Cooper Mode: on)
And strictly speaking, the key is always there (so it cannot be missing), it can only have an empty value.
(Full Sheldon Cooper Mode: off)

**If the application key is not set, your user sessions and other encrypted data will not be secure!**

#### Additional Configuration

> You may also want to configure a few additional components of Lychee in `.env`, such as:

Reviewer 2: I know that you haven't changed this line, but I wonder why it says "You may also ... such as: ... database ...". Why "may"? The database is something which definitely "must" be configured.

I have to admit that I have never read the install instructions thoroughly, maybe I should have.

Reviewer: > The database is something which definitely "must" be configured.

Nope. Composer installs the example into `.env` if it doesn't exist, and the example uses a functional sqlite database.

Reviewer 2: Thanks for clarifying that. I didn't remember that. But then there are some other things which strikes me as wrong.

Some lines below, the guide says

> if you copied the `.env.example` file.

IMHO, this doesn't make sense then, because the file is always copied and set up for the user even if the user does nothing.

Moreover, I still believe that the order in which the information is presented is wrong. The `.env` file should be the primary source of configuration not `config/app.php`. Hence, this should be presented in that order and not the way how it is presented now.

Reviewer: > IMHO, this doesn't make sense then, because the file is always copied and set up for the user even if the user does nothing.

It's copied **if** the user does nothing. If it exists, it's not copied. Though I agree that's almost an edge case.

> Moreover, I still believe that the order in which the information is presented is wrong. The `.env` file should be the primary source of configuration not `config/app.php`. Hence, this should be presented in that order and not the way how it is presented now.

I don't understand your objection here - That is how they are presented in this PR, `.env` and then `config/` afterwards?

Reviewer 2: Well, yes after I changed it :grinning: We had a concurrency problem here.

Reviewer: That makes sense.

- Cache
- Database
- Session

> They are documented in the file if you copied the `.env.example` file.

Reviewer 2: Why "if you copied the ... file"? That's not optional, is it? Maybe, it should rephrase more like "copy the file" followed by "make the necessary adjustments".

Reviewer 2: As I learned just recently, the file is always copied by Composer, see https://github.com/LycheeOrg/LycheeOrg.github.io/pull/56#discussion_r889675863. Hence, the condition does make even less sense than it has made before.

In theory, you are right. If users create the file manually, they file won't contain any comments. But I think, this is nothing we should consider, because this somehow holds for everything. We should assume that the user follows the configuration guide. Otherwise we would also need to write conditions like "file xyz contains comments unless you have deleted them". If users deliberately do something else than written in the guide, then they should know what they are doing and won't need the comments anyway. This is in particularly true for the `.env` file. If users managed to write their own, working `.env` file from scratch without using the template, then they should also know that they file won't contain any comments unless they have written their own comments.

Reviewer: Maybe
```suggestion
They are documented in the `.env.example` file, which may have been copied into your `.env`. They are also listed on the [configuration](configuration.html) page.
```
?

Reviewer: @nagmat84's last comment arrived while I was reviewing. I don't mind assuming the comments are there, but I think it makes sense to link to the configurations table added in this PR.

> Some advanced options are not defined there by default, but you can take a look at `config/app.php` to find such options. Usually, you don't have to change them and can go with the defaults. **If you don't know what they're doing, do not change them.**

Reviewer 2: Nice!

However, I wonder why it is restricted to `config/app.php`? The file `app.php` is probably the last one which needs adjustments. Other files like `database.php` and `mailer.php` are much more likely. So I would write it like "take a look into the `config/` directory.

Reviewer: I'm very in favour of complete examples (ideally with the default values - I suggested changing those to match elsewhere).

As a user, having to copy every directive is a pain, and then you'd probably want to add context in case you need to edit later.

- Database
- Session

They are documented in the `.env.example` file, which may have been copied into your `.env`. They are also listed on the [configuration](configuration.html) page.

##### Advanced Configuration (Dangerous)

Some advanced options cannot be configured through the `.env` file. If you really need to change them, look in the `config/` directory.

**If you don't know what they're doing, do not change them.**

## Web Server Configuration

### General

> Configure your web server to make the `public` directory accessible via a URL. This could be the root of your web server (short and convenient) or, if you prefer, a subfolder such as `photos`. The `index.php` in this directory serves as the interface for all HTTP requests to Lychee's API.

Reviewer 2: ```suggestion
Configure your web server to make the `public` directory accessible via a URL. This could be the root of your web server (short and convenient) or, if you prefer, a subpath such as `photos`. The `index.php` in this directory serves as the interface for all HTTP requests to Lychee's API.
```

URLs don't have folders or directories. File systems do. Technically, the thing between two slashes is called component, but I believe nobody will understand that term. But "subfolder" really confuses me.

Moreover, I believe we should mention that the `AP_URL` variable in the `.env` file must match the URL.

Reviewer 3: Thanks; I wasn't happy with the word "subfolder" when I was writing it but I couldn't come up with a better alternative at the time...

> > {note} Do not expose the top level directory with your web server! This would allow public access to internal files like the database and your `.env` which can leak your credentials!

Reviewer 2: ```suggestion
> {note} Do not expose the top level directory of Lychee with your web server! This would allow public access to internal files like the application code and the `.env` file which contains sensitive information like the database credentials!
```

Reviewer: It *could* expose the DB (for sqlite), but yeah, I wasn't convinced about mentioning that either.
